rskj-core/src/test/java/co/rsk/peg/federation/FederationStorageProviderImplTests.java

```diff
@@ -983,6 +985,109 @@ void getNextFederationCreationBlockHeight_afterRSKIP186AndNoValueInStorage_retur
 
     }
 
+    @Test
+    void getLastRetiredFederationP2SHScript_beforeRSKIP186_storageIsNotAccessedAndReturnsEmpty() {
+
+        // Arrange
+
+        ActivationConfig.ForBlock activations = mock(ActivationConfig.ForBlock.class);
+        when(activations.isActive(ConsensusRule.RSKIP186)).thenReturn(false);
+
+        StorageAccessor storageAccessor = new InMemoryStorage();
+        Script expectedScript = createFederationScript();
+        System.out.println(Arrays.toString(expectedScript.getProgram()));
+        byte[] serializedScript = BridgeSerializationUtils.serializeScript(expectedScript);
+        // Putting some value in the storage just to then assert that before fork, the storage won't be accessed.
+        storageAccessor.saveToRepository(LAST_RETIRED_FEDERATION_P2SH_SCRIPT_KEY.getKey(), serializedScript);
+
+        // Act
+
+        FederationStorageProvider federationStorageProvider = new FederationStorageProviderImpl(storageAccessor);
+
+        // Assert
+        assertEquals(Optional.empty(), federationStorageProvider.getLastRetiredFederationP2SHScript(activations));
+
+    }
+
+    @Test
+    void getLastRetiredFederationP2SHScript_afterRSKIP186_getsValueFromStorage() {
+
+        // Arrange
+
+        ActivationConfig.ForBlock activations = mock(ActivationConfig.ForBlock.class);
+        when(activations.isActive(ConsensusRule.RSKIP186)).thenReturn(true);
+
+        StorageAccessor storageAccessor = new InMemoryStorage();
+        Script expectedScript = createFederationScript();
+        byte[] serializedScript = BridgeSerializationUtils.serializeScript(expectedScript);
+
+        storageAccessor.saveToRepository(LAST_RETIRED_FEDERATION_P2SH_SCRIPT_KEY.getKey(), serializedScript);
+
+        // Act
+
+        FederationStorageProvider federationStorageProvider = new FederationStorageProviderImpl(storageAccessor);
+        Optional<Script> actualValueOptional = federationStorageProvider.getLastRetiredFederationP2SHScript(activations);
+
+        // Assert
+
+        assertTrue(actualValueOptional.isPresent());
+        assertEquals(expectedScript, actualValueOptional.get());
+
+        // Setting in storage a different value to assert that calling the method again should return cached value
+
+        storageAccessor.saveToRepository(LAST_RETIRED_FEDERATION_P2SH_SCRIPT_KEY.getKey(), expectedScript.getProgram());
+
+        Optional<Script> actualCachedValueOptional = federationStorageProvider.getLastRetiredFederationP2SHScript(activations);
```

Comment:
nit: this name sounds a bit weird, not sure if the `Optional` suffix adds value

Reply:
I agree

Reply:
Updated.